tool/src/org/antlr/v4/automata/ParserATNFactory.java

```diff
@@ -472,24 +472,28 @@ public Handle elemList(List<Handle> els) {
             if ( el.left.getNumberOfTransitions()==1 ) tr = el.left.transition(0);
             boolean isRuleTrans = tr instanceof RuleTransition;
             if ( el.left.getStateType() == ATNState.BASIC &&
+				el.right != null &&
 				el.right.getStateType()== ATNState.BASIC &&
-				tr!=null && (isRuleTrans && ((RuleTransition)tr).followState == el.right || tr.target == el.right) )
-			{
+				tr!=null && (isRuleTrans && ((RuleTransition)tr).followState == el.right || tr.target == el.right) ) {
 				// we can avoid epsilon edge to next el
-				if ( isRuleTrans ) ((RuleTransition)tr).followState = els.get(i+1).left;
-                else tr.target = els.get(i+1).left;
+				Handle handle = null;
+				if (i + 1 < els.size())
+					handle = els.get(i + 1);
+				if (handle != null) {
+					if (isRuleTrans)
+						((RuleTransition) tr).followState = handle.left;
+					else
+						tr.target = handle.left;
+				}
 				atn.removeState(el.right); // we skipped over this state
 			}
 			else { // need epsilon if previous block's right end node is complicated
 				epsilon(el.right, els.get(i+1).left);
 			}
 		}
 		Handle first = els.get(0);
-		Handle last = els.get(n -1);
-		if ( first==null || last==null ) {
-			g.tool.errMgr.toolError(ErrorType.INTERNAL_ERROR, "element list has first|last == null");
-		}
-		return new Handle(first.left, last.right);
+		Handle last = els.get(n - 1);
+		return new Handle(first != null ? first.left : null, last != null ? last.right : null);
```

Comment:
Seems dangerous to remove an error message

Reply:
If I restore the check, two errors will occur:

```
error(20): Test.g4:2:30: internal error: Rule error undefined 
error(20):  internal error: element list has first|last == null 
```

The message of internal error `element list has first|last == null` is duplicated, unclear for the end-user, and does not help to resolve an issue. The first message is much more informative.